apps/s_client.c

```diff
@@ -142,22 +142,24 @@ static unsigned int psk_client_cb(SSL *ssl, const char *hint, char *identity,
     int ret;
     long key_len;
     unsigned char *key;
+    char *used_hint = (char *)hint;
 
     if (c_debug)
         BIO_printf(bio_c_out, "psk_client_cb\n");
     if (!hint) {
         /* no ServerKeyExchange message */
         if (c_debug)
             BIO_printf(bio_c_out,
-                       "NULL received PSK identity hint, continuing anyway\n");
+                       "NULL received PSK identity hint, continuing with the default\n");
+        used_hint = psk_identity;
     } else if (c_debug) {
         BIO_printf(bio_c_out, "Received PSK identity hint '%s'\n", hint);
     }
 
     /*
      * lookup PSK identity and PSK key based on the given identity hint here
      */
-    ret = BIO_snprintf(identity, max_identity_len, "%s", psk_identity);
+    ret = BIO_snprintf(identity, max_identity_len, "%s", used_hint);
```

Comment:
This is not correct. The hint is a hint sent from the server to the client to help the client decide which identity to use. It is not typically supposed to be used *as* the identity itself.

From RFC4279:
> Both clients and servers may have pre-shared keys with several
   different parties.  The client indicates which key to use by
   including a "PSK identity" in the ClientKeyExchange message (note
   that unlike in [SHAREDKEYS], the session_id field in ClientHello
   message keeps its usual meaning).  To help the client in selecting
   which identity to use, the server can provide a "PSK identity hint"
   in the ServerKeyExchange message.